opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,139 +6,254 @@
-          .put("date", ExprCoreType.TIMESTAMP)
-          .put("ip", OPENSEARCH_IP)
-          .put("geo_point", OPENSEARCH_GEO_POINT)
-          .put("binary", OPENSEARCH_BINARY)
-          .build();
+  @EqualsAndHashCode.Exclude
+  protected MappingType mappingType;
 
-  /**
-   * Mapping from extra OpenSearch type name which may map to same ExprType as above.
-   */
-  private static final Map<String, ExprType> EXTRA_OPENSEARCH_TYPE_TO_EXPR_TYPE_MAPPING =
-      ImmutableMap.<String, ExprType>builder()
-          .put("half_float", ExprCoreType.FLOAT)
-          .put("scaled_float", ExprCoreType.DOUBLE)
-          .put("date_nanos", ExprCoreType.TIMESTAMP)
-          .build();
+  // resolved ExprCoreType
+  protected ExprCoreType exprCoreType;
 
   /**
-   * The mapping between Type and legacy JDBC type name.
+   * Get a simplified type {@link ExprCoreType} if possible.
+   * To avoid returning `UNKNOWN` for `OpenSearch*Type`s, e.g. for IP, returns itself.
+   * @return An {@link ExprType}.
    */
-  private static final Map<ExprType, String> LEGACY_TYPE_NAME_MAPPING =
-      new ImmutableMap.Builder<ExprType, String>()
-          .put(OPENSEARCH_TEXT, "text")
-          .put(OPENSEARCH_TEXT_KEYWORD, "text")
-          .build();
+  public ExprType getExprType() {
+    if (exprCoreType != ExprCoreType.UNKNOWN) {
+      return exprCoreType;
+    }
+    return this;
+  }
 
   /**
-   * Parent of current type.
+   * Simple instances of OpenSearchDataType are created once during entire SQL engine lifetime
+   * and cached there. This reduces memory usage and increases type comparison.
+   * Note: Types with non-empty fields and properties are not cached.
    */
-  private final List<ExprType> parents;
+  private static final Map<String, OpenSearchDataType> instances = new HashMap<>();
+
   /**
-   * JDBC type name.
+   * A constructor function which builds proper `OpenSearchDataType` for given mapping `Type`.
+   * @param mappingType A mapping type.
+   * @return An instance or inheritor of `OpenSearchDataType`.
    */
-  private final String jdbcType;
+  public static OpenSearchDataType of(MappingType mappingType) {
+    var res = instances.getOrDefault(mappingType.toString(), null);
+    if (res != null) {
+      return res;
+    }
+    ExprCoreType exprCoreType = mappingType.getExprCoreType();
+    if (exprCoreType == ExprCoreType.UNKNOWN) {
+      switch (mappingType) {
+        // TODO update these 2 below #1038 https://github.com/opensearch-project/sql/issues/1038
+        case Text: return OpenSearchTextType.of();
+        case GeoPoint: return OpenSearchGeoPointType.of();
+        case Binary: return OpenSearchBinaryType.of();
+        case Ip: return OpenSearchIpType.of();
+        default:
+          throw new IllegalArgumentException(mappingType.toString());
+      }
+    }
+    res = new OpenSearchDataType(mappingType);
+    res.exprCoreType = exprCoreType;
+    instances.put(mappingType.toString(), res);
```

Comment:
why not init the instances statically?

Reply:
Thank you for the idea! Fixed in be7e9c76f.